### DISCUSSION TITLE
Problem when developing multiple packages concurrently

### DISCUSSION THREAD

**Discussion by bilogic** [composer/composer#11958] [Q&A / Support]
```
project
├── composer.json
├── tests
├── app                             (A0)
│   ├── Facades
│   └── Http
│       └── Controllers
├── packages
│   └── mine
│       ├── package-1
│       │   ├── src                 (S1)
│       │   │   └── Facades     
│       │   ├── composer.json       (C1)
│       │   └── vendor              (V1)
│       └── package-2           
│           ├── composer.json       (C2)
│           ├── tests               (T2)
│           |   └── Unit
│           └── vendor              (V2)
│               └── mine
│                   └── package-1
├── vendor                          (V0)
│   ├── laravel
│   │   ├── framework
```

1. My folders are as above
2. I'm trying to develop package-1 and 2 oncurrently and package-2 requires code found in package-1 
3. Both packages have `autoload-dev` in their respective `composer.json`, so their `Tests` namespace only available in S1 and S2, not A0
    ```            
        "autoload-dev": {
            "psr-4": {
                "Mine\\Package-X\Tests\\": "tests/"
            }
        },
    ```
4. Problem: When I'm running PHPUnit tests for T2, they fail because they are using code from V2. How do I get them to use code in S1/V0 instead? And is there a way to do it without having to manipulate C2 manually between dev and production?

It took me quite a while before this problem became clear enough for me to explain, hope someone who has solved this before can share. Thank you!

**Comment by Seldaek** [comment]
I would say try https://github.com/franzliedke/studio or try [path repos](https://getcomposer.org/doc/05-repositories.md#path). Not sure what else we can do to help :)

**Comment by MatteoOreficeIT** [comment]
Hello everyones, I also would ask @Seldaek  on the same topic: 

> is composer designed to allow multiple require of `autoload.php` in the same runtime / execution flow?

What kind of issues could arise ? 

- build time conflict: because composer update / require is run in two separate times are random names generated in `vendor/composer/autoload*` guaranteed with respect to name collision?

- indirect dependencies conflict : the two composer.json have different roots but some indirect deps are there, is it possible to run into problems at runtime?

I discovered a particular layout in our company project where two different `autoload.php` are included during a regular API call from two indipendent `vendor` locations, and I am asking myself if it is undocumented and not recommended

I would appreciate a note in the doc about the design and if this feature is or isn't supported officially. 

> IMHO I would use a differente approach with path repo to obtain the same result

Thanks !!!

**Reply by bilogic** [reply]
Sounds quite different from the problem I faced in point 4, I don't think 1 piece of code should ever include 2 different `autoload.php`, it is almost certainly bound to face a namespace clash and other issues.

**Reply by MatteoOreficeIT** [reply]
I thought the same , but composer documentation doesn't mention any warning about it, and I usually stay away from undocumented features !

**Reply by bilogic** [reply]
Since it is a different problem, might be more fruitful for you to start new a conversation

**Reply by MatteoOreficeIT** [reply]
Yes thanks !
